86 update run scripts

crispy-wonton opened the pull request:
Fixes #86 

# Description

This PR changes the key run scripts in the repo pipeline so that they can all be run with the same inputs (preprocessed & deduplicated EPC dataset), allowing them to be run out of order. This is meant to simplify and clarify the running of the full pipeline.

NB: I have also left comments on the files themselves to help you through the review. See below a summary of the main and minor changes (not in any order of importance).

Main changes:
- `pipeline/README.md` - new instructions on how to run simplified pipeline.
- `pipeline/prepare_features/garden_size.py` - new function (taken from previously reviewed script) to deduplicate garden size)
- `pipeline/prepare_features/land_area.py` & `pipeline/prepare_features/number_of_households.py` - deleted. Functions refactored and moved to `pipeline/prepare_features/property_density.py`.
- `pipeline/prepare_features/output_areas.py` - new function to load and clean postcode to LSOA lookup. We had a function that did this already but it also loaded MSOA, RUC, LAD columns which are unnecessary for the use case in which this new function was applied. Since the ONS PD is so big, I thought it would be more efficient to have a separate smaller function, rather than process more columns and then drop them.
- `pipeline/prepare_features/property_density.py` - Refactoring to simplify existing property density code and combine Scotland with England and Wales property density df generation into a single function.
- `pipeline/prepare_features/protected_areas.py` - New function to generate single df of protected areas from Scotland (world heritage sites) and England and Wales (building conservation areas).
- `pipeline/run_scripts/run_add_features.py` & `pipeline/run_scripts/run_calculate_garden_size.py` - Uses new and refactored functions and script is refactored to take preprocessed deduplicated EPC as input.
- `pipeline/run_scripts/run_calculate_suitability.py` - `calculate_suitability.py` refactored into 3 files, including this one. The whole file is showing diffs because it is newly created, but this script contains code that has been previously reviewed. I have labelled parts that have changed since the last review of this code, the rest is unchanged since the last review.
- `pipeline/suitability/calculate_suitability.py` - refactored for various reasons as labelled on file


Minor changes:
- `getters/get_datasets.py` - update documentation of some getters for clarity on type of land area measurement
- `analysis/exploratory/aggregate_build_year_per_la.py`  & `analysis/exploratory/subset_epc_north_england.py` & `pipeline/run_scripts/run_generate_open_dataset.py` & `pipeline/run_scripts/run_process_garden_size.py`- update call to renamed function
- `pipeline/prepare_features/anchor_properties.py` &  `pipeline/prepare_features/grid_capacity.py` - update outputs of functions to `polars` dataframes so that outputs can be joined to main EPC dataframe in `run_add_features.py`
- `pipeline/prepare_features/lat_lon.py` - minor correction to code
- `pipeline/reweight_epc/prepare_sample.py` - correction to code
- `pipeline/prepare_features/off_gas.py` - standardise doc string and column name
- `pipeline/run_scripts/run_compute_epc_weights.py` - Updated to use new ONSPD load transform function from `output_areas.py`
- `pipeline/suitability/scoring.py` - moved scoring dicts to separate file. These are unchanged, just moved to their own file.
- `utils/save_utils.py` - Save util updates to allow saving `.parquet` or `.csv` files


# Instructions for Reviewer

In order to test the code in this PR you need to ...
There are instructions at the top of each run script for how to run the code, as well as instructions in the pipeline/README if you want to run any of them.
I have run the updated scripts and they seem to work for me.

Please pay special attention to ...
- the new pipeline/README is clear and understandable. I'm not sure about the table format, so let me know if it doesn't work. I wanted to create something that could be easily updated. Let me know if I should change the order or anything to make it more clear.
- New functions
- Refactoring of existing functions to ensure they produce the same output, especially refactoring of the functions to calculate suitability
- All run scripts have been appropriately refactored as required

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

crispy-wonton committed:
update generate_gdf_uprn_coords in lat_lon.py to use inputted x and y coord columns instead of hard-coding

crispy-wonton committed:
delete land_area.py and number_of_households.py and move&refactor functions into property_density.py
- add new function to generate property density for all of GB

crispy-wonton committed:
rename `off_gas` column created in functions in off_gas.py for consistency

crispy-wonton committed:
update output dfs in anchor_properties.py and grid_capacity.py to be polars DataFrames so they can be joined to EPC

crispy-wonton committed:
add new function to output_areas.py to generate dataset with just LSOA and postcode columns and rename existing function

crispy-wonton committed:
add `otherwise(None)` to add_col_property_type function in prepare_sample.py

crispy-wonton committed:
update documentation for getters in get_datasets.py

crispy-wonton committed:
update run_compute_epc_weights.py to add only LSOA col instead of also adding MSOA, RUC indicator, LAD

crispy-wonton committed:
update EPC cols used in calculate_suitability.py

crispy-wonton committed:
update run_add_features.py to take preprocessed dedupl EPC as input instead of weighted EPC
- refactor property density code so that England and Wales and Scotland property density can be added in one function
- add LSOA, MSOA, LAD, and RUC indicator cols to dataset
- select specific required from anchor_properties and grid_capacity dfs to create final clean dataset

crispy-wonton committed:
refactor calculate_suitability.py into function module, scoring.py contain score dicts, and run_calculate_suitability.py run script

crispy-wonton committed:
update save_utils.py to save different file types and update usages in exploratory notebooks and run scripts

crispy-wonton committed:
add new function to garden_size.py to deduplicate UPRNs

crispy-wonton committed:
add new function to protected_areas.py to generate full df with E/W/S protected areas (either building conservation or WHS)

crispy-wonton committed:
update run_add_features.py to use new function to generate protected areas for all nations

crispy-wonton committed:
add code to run_calculate_garden_size.py to generate two files, one with full dataset and other with deduplicated UPRNs

crispy-wonton committed:
update run_process_garden_size.py to use updated save_utils function

crispy-wonton committed:
update pipeline/README with new instructions to run the full pipeline

crispy-wonton committed:
correct error in pipeline/README table

crispy-wonton committed:
update run_calaculate_garden_size.py script to take preprocessed & deduplicated EPC
add try/except block for loading microsoft building footprint files

crispy-wonton commented on asf_heat_pump_suitability/getters/get_datasets.py:
```diff
@@ -274,7 +275,7 @@ def load_gdf_scotgov_data_zone_bounds() -> gpd.GeoDataFrame:
 
     Returns:
-        gpd.GeoDataFrame: boundary polygons and area data for 2011 Scottish Data Zones
+        gpd.GeoDataFrame: boundary polygons and area standard area measurment data for 2011 Scottish Data Zones
```
```suggestion
        gpd.GeoDataFrame: boundary polygons and area standard area measurement data for 2011 Scottish Data Zones
```

crispy-wonton commented on asf_heat_pump_suitability/pipeline/README.md:
Easier to review this file by clicking the three dots in the top right and selecting 'View file'.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/prepare_features/anchor_properties.py:
```diff
@@ -103,12 +102,12 @@ def load_gdf_and_process_poi() -> gpd.GeoDataFrame:
 
 
-def identify_anchor_properties_gdf() -> gpd.GeoDataFrame:
+def identify_anchor_properties_df() -> pl.DataFrame:
```
I changed the output of this function to a `polars.DataFrame` object so that it can be joined to the EPC dataset (which is also a `polars.DataFrame`) in the `run_add_features.py` script.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/prepare_features/garden_size.py:
```diff
@@ -126,3 +127,28 @@ def generate_gdf_garden_size(
     return gardens_gdf
+
+
+def deduplicate_df_garden_size(df: pl.DataFrame) -> pl.DataFrame:
```
Just the docstring needs reviewing here. The code has already been reviewed as it was moved from a script into this function without changing.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/prepare_features/grid_capacity.py:
```diff
@@ -239,7 +239,7 @@ def assess_heatpump_suitability(
 
 
-def calculate_grid_capacity() -> pd.DataFrame:
+def calculate_grid_capacity() -> pl.DataFrame:
```
I changed the output of this function to a `polars.DataFrame` object so that it can be joined to the EPC dataset (which is also a `polars.DataFrame`) in the `run_add_features.py` script.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/prepare_features/land_area.py:
Deleted this file, refactored the functions and moved them to `property_density.py`.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/prepare_features/number_of_households.py:
Deleted this file, refactored the functions and moved them to `property_density.py`.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/prepare_features/protected_areas.py:
Here I combined the functions to generate the `in_building_conservation_area` flag for UPRNs in England and Wales, and the `in_world_heritage_site` flag for UPRNs in Scotland. I combined the output into one single feature `in_protected_area` to make it easier to use in calculating suitability.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_garden_size.py:
```diff
@@ -112,9 +123,16 @@ def parse_arguments() -> argparse.Namespace:
-        building_footprints_gdf = building_footprint.transform_gdf_building_footprints(
-            building_file
-        )
+        try:
```
This `try`/`except` was added to deal with some of the Scottish `.shp` files which had some corrupted polygons.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_garden_size.py:
```diff
@@ -150,7 +169,13 @@ def parse_arguments() -> argparse.Namespace:
-    epc_gardens_df.to_parquet(args.save_as, engine="pyarrow")
+    save_utils.save_to_s3(epc_gardens_df, args.save_as)
+
+    logging.info("Deduplicating UPRNs that were matched to multiple gardens")
```
this deduplication was applied in a separate script previously (`run_process_garden_size.py`) but I wanted to put it together

crispy-wonton commented on asf_heat_pump_suitability/pipeline/suitability/calculate_suitability.py:
This file has been refactored into 3 files:
1. `pipeline/suitability/scoring.py` - scoring dicts
2. `pipeline/suitability/calculate_suitability.py` - functions used to calculate suitability
3. `pipeline/run_scripts/run_calculate_suitability.py` - script to calculate suitability

crispy-wonton commented on asf_heat_pump_suitability/pipeline/suitability/scoring.py:
These dicts are unchanged from their originals, they've just been moved to a new location. They don't need to be reviewed.

crispy-wonton committed:
add line to run_calculate_garden_size.py to remove dfs with len 0

crispy-wonton commented on asf_heat_pump_suitability/pipeline/prepare_features/lat_lon.py:
```diff
@@ -39,7 +39,7 @@ def generate_gdf_uprn_coords(
         usecols = ["*"]
     else:
-        for col in ["X_COORDINATE", "Y_COORDINATE"]:
+        for col in [x_col, y_col]:
```
This line still worked before, but this is more correct since we take these columns as inputs to the function.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/prepare_features/off_gas.py:
Here, I've just standardised the docstring a bit and also lower-cased the resulting `off_gas` column name for consistency with the other features.

lizgzil replied:
makes sense!

crispy-wonton commented on asf_heat_pump_suitability/pipeline/prepare_features/output_areas.py:
New function to load and clean postcode to LSOA lookup. We had a function that did this already but it also loaded MSOA, RUC, LAD columns which are unnecessary for the use case in which this new function was applied. Since the ONS PD is so big, I thought it would be more efficient to have a separate smaller function, rather than process more columns and then drop them.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/prepare_features/property_density.py:
Refactoring to simplify existing property density code and combine Scotland with England and Wales property density df generation into a single function.

lizgzil replied:
this looks good and clearer. I was confused where some of the cleaning steps went though - don't you still have the division by 0 issue?

crispy-wonton replied:
See comment below: https://github.com/nestauk/asf_heat_pump_suitability/pull/93/files/8df29d6f5984958a87c6e30ae3ccf119c2bde284#r1892836196

crispy-wonton commented on asf_heat_pump_suitability/pipeline/prepare_features/protected_areas.py:
New function to generate single df of protected areas from Scotland (world heritage sites) and England and Wales (building conservation areas).

lizgzil replied:
much clearer :)

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:
Uses new and refactored functions and script is refactored to take preprocessed deduplicated EPC as input.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_garden_size.py:
Uses new and refactored functions and script is refactored to take preprocessed deduplicated EPC as input.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_suitability.py:
`calculate_suitability.py` refactored into 3 files, including this one. The whole file is showing diffs because it is newly created, but this script contains code that has been previously reviewed. I have labelled parts that have changed since the last review of this code, the rest is unchanged since the last review.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_suitability.py:
```diff
@@ -0,0 +1,150 @@
+from asf_heat_pump_suitability.pipeline.suitability import calculate_suitability
+
+
+def parse_arguments():
```
Parser function updated since last review to take the paths of the different input files.

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_suitability.py:
```diff
@@ -0,0 +1,150 @@
+
+
+if __name__ == "__main__":
+    args = parse_arguments()
+    y = args.year
+    q = args.quarter
+
+    logging.info("Loading EPC data with features")
+    epc_df = pl.read_parquet(path=args.features)
+    logging.info("Loading garden size estimates")
+    gardens = pl.read_parquet(path=args.gardens)
+    logging.info("Loading weights")
+    weights = pl.read_parquet(path=args.weights)
+
+    logging.info("Joining EPC features data with garden size estimates and weights")
+    epc_df = epc_df.join(gardens, how="left", on="UPRN")
+    epc_df = epc_df.join(weights, how="left", on="UPRN")
+    epc_df = epc_df.with_columns(
+        pl.col("garden_area_m2")
+        .fill_null(pl.col("msoa_avg_outdoor_space_m2"))
+        .alias("garden_area_m2")
+    ).drop("msoa_avg_outdoor_space_m2")
```
lines 62-80 are new

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_suitability.py:
```diff
@@ -0,0 +1,150 @@
+
+
+if __name__ == "__main__":
+    args = parse_arguments()
+    y = args.year
+    q = args.quarter
+
+    logging.info("Loading EPC data with features")
+    epc_df = pl.read_parquet(path=args.features)
+    logging.info("Loading garden size estimates")
+    gardens = pl.read_parquet(path=args.gardens)
+    logging.info("Loading weights")
+    weights = pl.read_parquet(path=args.weights)
+
+    logging.info("Joining EPC features data with garden size estimates and weights")
+    epc_df = epc_df.join(gardens, how="left", on="UPRN")
+    epc_df = epc_df.join(weights, how="left", on="UPRN")
+    epc_df = epc_df.with_columns(
+        pl.col("garden_area_m2")
+        .fill_null(pl.col("msoa_avg_outdoor_space_m2"))
+        .alias("garden_area_m2")
+    ).drop("msoa_avg_outdoor_space_m2")
+
+    logging.info("Filtering EPC data to rows with n_features >= minimum threshold")
+    features = [
+        "CURRENT_ENERGY_RATING",
+        "property_type",
+        "ruc_two_fold",
+        "off_gas",
+        "listed_building",
+        "in_protected_area",
+        "garden_area_m2",
+        "households_per_km2",
+        "has_anchor_properties",
+        "heatpump_installation_percentage",
+    ]
+    epc_df = calculate_suitability.filter_df_minimum_features(epc_df, features=features)
```
lines 62-95 are new (although the functions used already existed)

lizgzil replied:
nice and clear!

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_suitability.py:
```diff
@@ -0,0 +1,150 @@
+        epc_df = epc_df.join(score_df, on="UPRN", how="left")
+
+    save_as = f"s3://asf-heat-pump-suitability/outputs/{y}{q}/{datetime.today().strftime('%Y%m%d')}_{y}_Q{q}_heat_pump_suitability_per_property.parquet"
+    save_utils.save_to_s3(epc_df, save_as)
```
This save util has been updated

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_suitability.py:
```diff
@@ -0,0 +1,150 @@
+
+    logging.info("Saving LSOA heat pump suitability scores")
+    save_as = f"s3://asf-heat-pump-suitability/outputs/{y}{q}/{datetime.today().strftime('%Y%m%d')}_{y}_Q{q}_heat_pump_suitability_per_lsoa"
+    save_utils.save_to_s3(suitability_df, f"{save_as}.parquet")
+    save_utils.save_to_s3(suitability_df, f"{save_as}.csv")
```
These save utils have been updated

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py:
Updated to use new ONSPD load transform function from `output_areas.py`

crispy-wonton commented on asf_heat_pump_suitability/pipeline/suitability/calculate_suitability.py:
Refactoring to:
- Move scoring dicts into separate `scoring.py` file
- Ensure most up-to-date feature names are used, including Scotland
- Move script portion of file to `run_calculate_suitability.py`
- General refactoring in some functions where required.

lizgzil replied:
Makes a lot of sense!

crispy-wonton commented on asf_heat_pump_suitability/pipeline/suitability/scoring.py:
These are unchanged, just moved to their own file.

crispy-wonton commented on asf_heat_pump_suitability/utils/save_utils.py:
Save util updates to allow saving `.parquet` or `.csv` files

lizgzil commented on asf_heat_pump_suitability/pipeline/prepare_features/garden_size.py:
```diff
@@ -126,3 +127,28 @@ def generate_gdf_garden_size(
+
+def deduplicate_df_garden_size(df: pl.DataFrame) -> pl.DataFrame:
+    """
+    Deduplicate UPRNs matched to multiple gardens them by taking the average size of the multiple gardens (for gardens
```
```suggestion
    Deduplicate UPRNs matched to multiple gardens by taking the average size of the multiple gardens (for gardens
```

lizgzil commented on asf_heat_pump_suitability/pipeline/prepare_features/property_density.py:
```diff
@@ -2,67 +2,76 @@
     """
-    # Fill None values with a default value (e.g., 0)
-    enhanced_epc_df = enhanced_epc_df.with_columns(
-        [
-            pl.col("Number of households 2021")
-            .fill_null(0)
-            .alias("Number of households 2021"),
-            pl.col("Land Count (Area in KM2)")
-            .fill_null(0)
-            .alias("Land Count (Area in KM2)"),  # Fill with 1 to avoid division by zero
```
how come you didn't need to do this in the new version?

crispy-wonton replied:
It's no longer needed because when calculating property density, instead of calculating for the whole EPC dataset, I only calculate it where sufficient data is available - i.e. both household count and land area data is available. Household counts is joined to area data with an `inner` join. The minimum value for land area is 0.0097 km2, so we don't end up dividing anything by 0.

The resulting property density df is then left joined to the full EPC dataset. So where property density information is not available, there will be nulls in the final EPC dataset in this column.

lizgzil commented on asf_heat_pump_suitability/pipeline/prepare_features/protected_areas.py:
```diff
@@ -6,6 +6,27 @@
+        {"in_world_heritage_site": "in_protected_area"}
+    )
+
+    return pl.concat([ew_df, s_df])
```
are you sure there is no overlap where some UPRNs are in both? You might need to make sure that the UPRNs aren't repeated when you do `pl.concat([ew_df, s_df]))`

crispy-wonton replied:
Uh oh, you have found a bug!! Thank you! :D 

This is definitely a problem and needs to be fixed. Here we have ended up with not only duplicate UPRNs, but duplicated UPRNs with opposing `in_protected_areas` values!

This is because when we check if a UPRN is in a World Heritage Site in Scotland, any that aren't in a WHS are flagged as `False` (including those not even in Scotland). So if a property in England or Wales is in a conservation area, it will receive both a `True` and `False` flag!

crispy-wonton replied:
This should be fixed now with the new code in `generate_df_uprn_in_whs` function below. Now, we filter EPC data to Scotland properties only before checking if they are in a WHS. Then we concat only scottish data to england and wales data (which should just be england and wales due to the fact we are joining with conservation area data with an `inner` join). I checked the resulting df, and there are now no duplicated UPRNs in the combined dataset.

lizgzil commented on asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:
```diff
@@ -76,28 +81,48 @@ def parse_arguments() -> argparse.Namespace:
-    logging.info("Adding LAD code with geospatial join")
-    uprn_lad_df = output_areas.sjoin_df_uprn_lad_code(epc_gdf)
-    epc_df = epc_df.drop("lad_code").join(uprn_lad_df, how="left", on="UPRN")
+    # TODO this is far too slow and is only used to correct some EPC records with incorrect postcodes which get joined to the wrong LSOA/MSOA
+    # TODO can we filter the dataset somehow and only do the join with incorrect postcodes?
+    # # Replace `lad_code` from postcode with `lad_code` from geospatial join and postcode
+    # logging.info("Adding LAD code with geospatial join")
+    # uprn_lad_df = output_areas.sjoin_df_uprn_lad_code(epc_gdf)
+    # epc_df = epc_df.drop("lad_code").join(uprn_lad_df, how="left", on="UPRN")
```
was this supposed to be fixed in this PR?

lizgzil replied:
Something that might be quicker is to do the sjoin in batches in `sjoin_df_uprn_lad_code`.

```
from tqdm import tqdm
import pandas as pd

def list_chunks(orig_list: list, chunk_size: int = 100):
    """Chunks list into batches of a specified chunk_size."""
    for i in range(0, len(orig_list), chunk_size):
        yield orig_list[i : i + chunk_size]
        

def sjoin_df_uprn_lad_code(gdf: gpd.GeoDataFrame) -> pl.DataFrame:
    """
    Geospatial join between UPRNs with x,y coordinates and local authority (LAD) boundaries to match UPRNs with the code for
    the local authority they are located in. Null LAD codes are filled with LAD codes matched to UPRN on postcode.

    Args:
        gdf (gpd.GeoDataFrame): dataframe with point geometries per UPRN in BNG, and LAD code from postcode

    Returns:
        pl.DataFrame: UPRNs with matched local authority code
    """
    lad_bounds_gdf = get_datasets.load_gdf_ons_council_bounds(
        columns=["LAD23CD", "geometry"]
    )
    df_a1 = pd.DataFrame()
    for chunk_gdf in tqdm(list_chunks(gdf, chunk_size = 10000)):
        chunk_gdf_result = chunk_gdf.sjoin(lad_bounds_gdf, how="left", predicate="intersects")
        df_a1 = pd.concat([df_a1, chunk_gdf_result])
    gdf = df_a1
    gdf["lad_code"] = gdf["LAD23CD"].fillna(gdf["lad_code"])
    return pl.from_pandas(gdf[["UPRN", "lad_code"]])
        
```

excuse my variable namings.

Investigating different `chunk_size`, I got the following run times per iteration:

    # 50k: 34.52s/it, 33.90s/it, 33.72s/it
    # 10k: 9.13s/it, 6.87s/it, 6.99s/it, 7.08s/it, 6.90s/it
    # 1k: 1.46it/s, 1.59, 1.41, 1.39, 1.43, 1.46, 1.40

crispy-wonton replied:
Thanks so much for writing this @lizgzil ! This wasn't something that I was planning to address in this PR because I believe it's an edge case problem that only affects a few of the EPC records*.

Because these lines of code add about 1hr+ to the run time and don't add enough value, I commented them out for the time being. 

*The original edge case was found when we were adding the conservation area feature based on LAD code and noticed that a single EPC record had an erroneous postcode and therefore had been attributed to the wrong local authority, thus being incorrectly assigned a flag that said it was in an LA without conservation area data while also being flagged as in a conservation area.

crispy-wonton replied:
I have added issue #95 with your suggestion to address this separately @lizgzil .

lizgzil commented on asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:
```diff
@@ -108,58 +133,30 @@ def parse_arguments() -> argparse.Namespace:
-    epc_df = epc_df.with_columns(
-        pl.col("households_per_km2").fill_null(pl.col("households_per_km2_right"))
-    ).drop("households_per_km2_right")
+    lsoa_density_df = property_density.generate_df_property_density()
```
much nicer! 😄

lizgzil commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_suitability.py:
```diff
@@ -0,0 +1,150 @@
+                lsoa_df, lsoa_code
+            )
+        )
+    # Must have at least 15 properties to be included in score
```
Is this is in a readme somewhere so we don't forget this filtering step?

lizgzil commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_suitability.py:
```diff
@@ -0,0 +1,150 @@
+    suitability_df = pl.DataFrame(weighted_scores).filter(pl.col("n_properties") >= 15)
+    suitability_df = suitability_df.with_columns(pl.col(pl.Float64).round(3))
+
+    # TODO add Data Zone names
```
is this in an issue? (don't think I see it)

crispy-wonton replied:
Good flag, new issue created here : https://github.com/nestauk/asf_heat_pump_suitability/issues/94

lizgzil commented on asf_heat_pump_suitability/pipeline/suitability/calculate_suitability.py:
```diff
@@ -256,46 +56,47 @@ def compute_df_total_score_per_epc(
         .alias("anchor_properties_score"),
         pl.when(
-            pl.col("heatpump_installation_percentage") > grid_installation_threshold
+            pl.col("heatpump_installation_percentage") >= grid_installation_threshold
         )
-        .then(epc_threshold_scores.get(tech_type))
+        .then(scoring.epc_threshold_scores.get(tech_type))
         .alias("grid_capacity_score"),
```
I think this need updating since it uses the `epc_threshold_scores` scores (even though they are the same).

So let's have another scores dict in scoring with:
```
grid_capacity_scores = {
    "ASHP_S": 1,
    "ASHP_N": 0,
    "GSHP_S": 1,
    "GSHP_N": 0,
    "SGL_S": 1,
    "SGL_N": 0,
    "HN_S": 0,
    "HN_N": 0,
}
```

Also I like the idea of the 'grid_capacity_score' being relative to the actual `heatpump_installation_percentage` value, rather than a binary based off a threshold. So perhaps this is better:
```
(pl.col("heatpump_installation_percentage")/100)*scoring.grid_capacity_scores.get(tech_type).alias("grid_capacity_score")
```
[if that is the right polars syntax!]

What do you think?

crispy-wonton replied:
Thanks @lizgzil ! Great spot on the `epc_threshold_scores`. Weird that we didn't use the `grid_capacity_scores` dict before because it did already exist so not sure what happened! I've updated it now.

I've also updated the scoring calculation based on your suggestion which I think is a good idea because it gives more nuance in the final scores. So now, if the `heatpump_installation_percentage` is not null, then the score will be multiplied by the %. For this reason, we no longer require the `grid_installation_threshold` arg so I have removed it.

lizgzil commented on asf_heat_pump_suitability/pipeline/suitability/scoring.py:
```diff
@@ -0,0 +1,131 @@
+    "SGL_N": 2,
+    "HN_S": 2,
+    "HN_N": 2,
+}
```
To add 

```
grid_capacity_scores = {
    "ASHP_S": 1,
    "ASHP_N": 0,
    "GSHP_S": 1,
    "GSHP_N": 0,
    "SGL_S": 1,
    "SGL_N": 0,
    "HN_S": 0,
    "HN_N": 0,
}
```

crispy-wonton replied:
I realised it's actually already in there, just wasn't used for some reason! Good catch thank you!

lizgzil commented on asf_heat_pump_suitability/pipeline/README.md:
```diff
@@ -11,98 +11,40 @@ asf_heat_pump_suitability/pipeline
+To weight the EPC data, add new features, and estimate garden size of properties in preparation for calculating suitability, you can run the
+following files in any order. All scripts take the preprocessed and deduplicated EPC dataset (output from `asf-daps`) in
+parquet file format as input. Ensure you set the `--year` and `--quarter` arguments to correspond to those of the EPC
+dataset when running each script.See the script `.py` files for more detailed running instructions.
```
I know it will become outdated, so will need updating, but could you include here the latest arguments you use in `--epc [path/to/EPC] -y [YYYY] -q [N]` it's never clear to me which ones to run!

crispy-wonton replied:
No problem, I think that's a good idea! :D

lizgzil commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_suitability.py:
```diff
@@ -0,0 +1,150 @@
+"""
+Functions to calculate suitability of different HP technologies.
```
This could do with the instructions of what to run like you've done with the others

lizgzil reviewed (commented):
This looks really great @crispy-wonton  - so much clearer! Well done - must have been a lot of gnarly refactoring! 🎉 

I added some comments/questions.

I was wondering when reading through everything again whether all the suitability scoring assumptions are captured in a readme somewhere? - thinking of the various filters which decide whether a LSOA is included - e.g. >=15 properties, >=4 features, min proportion of properties in lsoa with weights.

I also had a comment on how we utilise the grid capacity feature.

I do find the readme tables slightly hard to parse since they render in a way that means I have to scroll across to read them all - so can't see them at a glance. But the info is there, so don't feel like you need to change these if you find them clearer this way.

I tried running `python -i asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py --epc s3://asf-daps/lakehouse/processed/epc/old/deduplicated/processed_dedupl-0.parquet -y 2023 -q 4` and got an error at the `grid_capacities =grid_capacity.calculate_grid_capacity().select(["lsoa", "heatpump_installation_percentage"])` line. Seems to be something to do with the loading?

```
File "/Users/elizabethgallagher/Code/asf_heat_pump_suitability/asf_heat_pump_suitability/getters/get_dno_datasets.py", line 289, in generate_ssen_gdf
    sepd_shape = gpd.read_file(config["data_source"]["S_SSEN_sepd_bounds"])
  File "/Users/elizabethgallagher/miniconda3/envs/asf_heat_pump_suitability/lib/python3.10/site-packages/geopandas/io/file.py", line 294, in _read_file
    return _read_file_pyogrio(
  File "/Users/elizabethgallagher/miniconda3/envs/asf_heat_pump_suitability/lib/python3.10/site-packages/geopandas/io/file.py", line 547, in _read_file_pyogrio
    return pyogrio.read_dataframe(path_or_bytes, bbox=bbox, **kwargs)
  File "/Users/elizabethgallagher/miniconda3/envs/asf_heat_pump_suitability/lib/python3.10/site-packages/pyogrio/geopandas.py", line 261, in read_dataframe
    result = read_func(
  File "/Users/elizabethgallagher/miniconda3/envs/asf_heat_pump_suitability/lib/python3.10/site-packages/pyogrio/raw.py", line 196, in read
    return ogr_read(
  File "pyogrio/_io.pyx", line 1239, in pyogrio._io.ogr_read
  File "pyogrio/_io.pyx", line 219, in pyogrio._io.ogr_open
pyogrio.errors.DataSourceError: 404: Send failure: Broken pipe
```

Perhaps something to do with a package version I have? I need to log off soon, but I'll investigate more tomorrow - but wanted to get my comments out now as they are.

lizgzil commented on asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:
```diff
@@ -169,16 +166,18 @@ def parse_arguments() -> argparse.Namespace:
+    grid_capacities = grid_capacity.calculate_grid_capacity().select(
+        ["lsoa", "heatpump_installation_percentage"]
+    )
     epc_df = epc_df.join(grid_capacities, how="left", on="lsoa")
```
```suggestion
    epc_df = epc_df.join(grid_capacities, how="left", on="lsoa").with_columns(
    pl.col("heatpump_installation_percentage").fill_null(False)
)
```

crispy-wonton replied:
I think the nulls in `heat_pump_installation_percentage` and `has_anchor_property` are true nulls. Meaning I think we are missing the data to say what their HP installation percentage is and whether or not they have anchor properties. Therefore I think they should remain as nulls rather than filling with False.

@shabrf is this correct?

lizgzil replied:
ah I see!

shabrf replied:
For heat pump installations the null values would be true nulls as these appear when we don't have geo data for the substation distribution area which covers the LSOA, so the capacity available to that LSOA is unknown. For anchor properties, we can assume the the points of interest dataset has good coverage so null values here would indicate that the LSOA doesn't contain any of the specified anchor properties within it, so I think we should fill these nulls with False.

crispy-wonton replied:
Thanks @shabrf for your helpful comment! I've updated the code to fill `has_anchor_property` nulls with `False`.

lizgzil commented on asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py:
```diff
@@ -169,16 +166,18 @@ def parse_arguments() -> argparse.Namespace:
+    anchor_properties_df = anchor_properties.identify_anchor_properties_df().select(
+        ["lsoa", "has_anchor_property"]
+    )
     epc_df = epc_df.join(anchor_properties_df, how="left", on="lsoa")
```
```suggestion
    epc_df = epc_df.join(anchor_properties_df, how="left", on="lsoa").with_columns(
    pl.col("has_anchor_property").fill_null(False)
)
```

crispy-wonton replied:
As above https://github.com/nestauk/asf_heat_pump_suitability/pull/93/files#r1894150275

lizgzil reviewed (commented):
@crispy-wonton  - when I ran it again today the `grid_capacity.calculate_grid_capacity()` issue I had yesterday didn't come up again - so must have just been an internet connection thing or something like that! 😌 

I ran `run_add_features` in full (apart from saving) and it all ran, and the final epc_df looked good to me - just noticed some nulls that could be turned to False 👍

crispy-wonton committed:
Update asf_heat_pump_suitability/getters/get_datasets.py

crispy-wonton committed:
Correct typo in asf_heat_pump_suitability/pipeline/prepare_features/garden_size.py

Co-authored-by: Liz G <elizabeth.gallagher@nesta.org.uk>

crispy-wonton commented on asf_heat_pump_suitability/pipeline/prepare_features/property_density.py:
```diff
@@ -2,67 +2,76 @@
-    pl.DataFrame: The modified DataFrame.
+        pl.DataFrame: households per km2 per LSOA in England and Wales
     """
+    households_df = load_transform_df_n_households_ew()
+    area_df = load_transform_df_land_area_ew()
+    df = households_df.join(area_df, how="inner", on="lsoa")
```
@lizgzil here is the inner join between households and area which means we don't end up with any rows with 0 value for area. Therefore, we don't need the code to prevent zero divisions.

crispy-wonton committed:
fix bug in protected_areas.py when generating df of protected areas

crispy-wonton committed:
update pipeline/README with example args

crispy-wonton committed:
update docstrings of run scripts to change -q [N] to -q [Q] and add run instructions to docstring of run_calculate_suitability.py

crispy-wonton committed:
update main README with details on 50% threshold of weighted properties for use of weights and data required for at least 15 properties
- correct typo in pipeline/README
- correct notation in run_calculate_suitability.py

crispy-wonton committed:
Update README.md formatting

crispy-wonton commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_suitability.py:
```diff
@@ -0,0 +1,159 @@
+                lsoa_df, lsoa_code
+            )
+        )
+    # Must have at least 15 properties to be included in final dataset
```
@lizgzil I have now added a note about this to the main `README` based on your comment. Please check you're happy with it :)

crispy-wonton committed:
update protected_areas.py to filter to Scottish properties only before checking if in WHS - this solves the issue of UPRNs being duplicated when concatenating england, wales, and scotland

crispy-wonton committed:
fix saving to csv in anchor_properties.py and grid_capacity.py scripts after changing output df from pandas to polars

crispy-wonton committed:
update calculate_suitability.py to calculate `grid_capacity_score` relative to the actual heatpump_installation_percentage value, rather than a binary based off a threshold

crispy-wonton committed:
update README to include info about minimum 4 features required to calculate suitability score

crispy-wonton commented:
@lizgzil thank you for your review and comments.

I have addressed your comments and answered your questions. I've made 3 new issues based on your comments to address separately: #94 #95 #96 

I have fixed the bug you found in `protected_areas.py`, updated the READMEs as requested, and changed the way we use the grid capacity feature for scoring.

I left a comment about filling NaN values with `False`. I think it's correct that we leave them as NaN values but have requested @shabrf 's input to double check this.

lizgzil commented on asf_heat_pump_suitability/pipeline/README.md:
```diff
@@ -23,9 +23,15 @@ asf_heat_pump_suitability/pipeline
+- `--epc s3://asf-daps/lakehouse/processed/epc/old/deduplicated/processed_dedupl-0.parquet`
+- `--year 2023`
+- `--quarter 4`
+
```
thanks!

lizgzil commented on README.md:
```diff
@@ -61,7 +61,16 @@ loops (SGLs).
+
+If the threshold is not met for a given LSOA, suitability scores for that LSOA will be unweighted and labelled as such.
+Unweighted scores may not accurately represent the suitability of an LSOA for a given heating technology as a whole and
+should therefore be interpreted with caution.
```
thanks! this is really useful to have made a note of

lizgzil reviewed (dismissed):
Thanks for all this @crispy-wonton - looks great!

crispy-wonton committed:
fill nulls with False in `has_anchor_property` column in `run_add_features.py`

crispy-wonton committed:
save full augmented EPC dataset from `run_calculate_suitability.py`

lizgzil commented on asf_heat_pump_suitability/pipeline/run_scripts/run_calculate_suitability.py:
```diff
@@ -0,0 +1,159 @@
+    for tech_type in tech_types:
+        logging.info(f"Calculating suitability scores for tech type: {tech_type}")
+        epc_scores_df = calculate_suitability.compute_df_avg_score_per_epc(
+            epc_df, tech_type
```
just seen that this has a default `density_threshold=100`. Is this as it should be? I got confused since the`compute_df_total_score_per_epc` function has a default of `density_threshold=60`.

crispy-wonton replied:
how the fweakin heck did you catch this! GREAT spot. I don't know which we should go with but they should be the same. I will investigate further.

crispy-wonton replied:
60 is the latest number we chose.
 
I found this comment in a slack thread for the reasoning but I don't fully know what it's referring to: "use a threshold of 60 households per km2 (found after fitting a regression for the data points with values of <100)".

crispy-wonton replied:
Corrected per your comment.

crispy-wonton committed:
correct threshold in `compute_df_avg_score_per_epc` function in calculate_suitability.py`

crispy-wonton committed:
Merge branch 'dev' into 86_update_run_scripts
